File: fastapi_migration/backend/app/schemas/analysis.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Union
from datetime import datetime

# Analysis request schemas
class AnalysisRequest(BaseModel):
    file_hash: Optional[str] = None
    analysis_id: Optional[str] = None
    raw_snp_data: Optional[List[Dict[str, Any]]] = None
    force_refresh: bool = False

# Analysis response schemas
class AnalysisResponse(BaseModel):
    analysis_id: str
    status: str
    message: str
    snp_count: int = 0
    file_hash: Optional[str] = None
    processing_time: Optional[float] = None
    cached: bool = False

class AnalysisResult(BaseModel):
    analysis_id: str
    created_at: Union[datetime, str]
    file_hash: Optional[str] = None
    data: Dict[str, Any]
    
    class Config:
        from_attributes = True

class AnalysisSummary(BaseModel):
    analysis_id: str
    file_hash: Optional[str] = None
    created_at: Union[datetime, str]
    status: str
    snp_count: int = 0
    
    class Config:
        from_attributes = True
        
class AnalysisList(BaseModel):
    items: List[AnalysisSummary]
    count: int

# Genetic structure schemas
class GeneticCharacteristic(BaseModel):
    name: str
    description: str
    effect_direction: Optional[str] = None
    evidence_strength: Optional[str] = None

class BeneficialIngredient(BaseModel):
    ingredient_name: str
    ingredient_mechanism: Optional[str] = None
    benefit_mechanism: str
    recommendation_strength: Optional[str] = None
    evidence_level: Optional[str] = None

class IngredientCaution(BaseModel):
    ingredient_name: str
    risk_mechanism: str
    alternative_ingredients: Optional[str] = None

class IngredientRecommendations(BaseModel):
    prioritize: List[BeneficialIngredient] = []
    caution: List[IngredientCaution] = []

class GeneticMutation(BaseModel):
    gene: str
    rsid: str
    allele1: str
    allele2: str
    risk_allele: str
    effect: str
    evidence_strength: Optional[str] = None
    category: Optional[str] = None
    characteristics: List[GeneticCharacteristic] = []

class AnalysisData(BaseModel):
    mutations: List[GeneticMutation] = []
    ingredient_recommendations: IngredientRecommendations
    summary: Optional[str] = None

File: zando/frontend/src/types/analysis.ts
/**
 * Types related to genetic analysis
 */

// Genetic traits and characteristics
export interface GeneticCharacteristic {
  name: string;
  description: string;
  effect_direction?: string;
  evidence_strength?: string;
}

// Mutation for the GeneCard component
export interface Mutation {
  rsid: string;
  gene_name: string;
  allele: string;
  effect: string;
  evidence_level: string;
  category?: string;
}

export interface GeneticMutation {
  gene: string;
  rsid: string;
  allele1: string;
  allele2: string;
  risk_allele: string;
  effect: string;
  evidence_strength?: string;
  category?: string;
  characteristics: GeneticCharacteristic[];
}

// Ingredient recommendations
export interface IngredientRecommendation {
  name: string;
  benefit?: string;
  caution?: string;
  genes?: string[];
  evidence_level?: string;
}

export interface BeneficialIngredient {
  ingredient_name: string;
  ingredient_mechanism?: string;
  benefit_mechanism: string;
  recommendation_strength?: string;
  evidence_level?: string;
}

export interface IngredientCaution {
  ingredient_name: string;
  risk_mechanism: string;
  alternative_ingredients?: string;
}

// Analysis data structure
export interface AnalysisData {
  mutations: Mutation[];
  ingredient_recommendations: {
    beneficial: IngredientRecommendation[];
    cautionary: IngredientRecommendation[];
  };
  summary?: string;
}

// API Request and Response types
export interface AnalysisRequest {
  file_hash?: string;
  analysis_id?: string;
  raw_snp_data?: any[];
  force_refresh?: boolean;
}

export interface AnalysisResponse {
  analysis_id: string;
  status: string;
  message: string;
  snp_count: number;
  file_hash?: string;
  processing_time?: number;
  cached: boolean;
}

export interface AnalysisResult {
  analysis_id: string;
  created_at: string;
  file_hash?: string;
  data: AnalysisData;
}

// Analysis state for the application
export interface AnalysisState {
  fileHash?: string;
  fileName?: string;
  analysisId?: string;
  snpCount?: number;
  status: 'idle' | 'uploaded' | 'analyzing' | 'analyzed' | 'error';
  error?: string;
  data?: AnalysisData;
}